python/pyarrow/__init__.py

```diff
@@ -281,6 +281,13 @@ def print_entry(label, value):
 
 import pyarrow.types as types
 
+try:
+    # Try importing the cuda module to ensure libarrow_cuda gets loaded
+    # to register the CUDA device for the C Data Interface import
+    import pyarrow.cuda
+except ImportError:
```

Comment:
I _think_ that this is required to get this working, but it's not actually tested right now. That's because the current tests use `pyarrow.cuda` anyway to create the test data. 
(to properly test it, we would need to use another library such as nanoarrow to create the data and then import in pyarrow without direct usage of `pyarrow.cuda`)

If we worry about import time that is being added here, I could also move this import to only do this on-demand at first usage of one of the `import_from_c_device` methods.

Reply:
Doing this on-demand, and/or from the tests sounds better to me.

Reply:
Based on a quick `python -X importtime` check, importing cuda here adds between 2 and 3% to the `pyarrow` import time

Reply:
>  and/or from the tests sounds better to me.

I think we should ensure this works out the box for users, without requiring them to first import `pyarrow.cuda` (they might not necessarily need anything from that module)?

Reply:
> Based on a quick `python -X importtime` check, importing cuda here adds between 2 and 3% to the `pyarrow` import time

Yes, I would be more worried by the potential implications of brinding CUDA into the process, though that's totally unsubstantiated.

Reply:
> I would be more worried by the potential implications of bringing CUDA into the process, though that's totally unsubstantiated.

For example worry about the memory usage increase (in case it wasn't actually needed)?

> Ok, so let's do it on demand if required?

You mean actually first check the device type in the struct?

Reply:
> For example worry about the memory usage increase (in case it wasn't actually needed)?

Yes.

> You mean actually first check the device type in the struct?

If easily possible, otherwise just everytime some device array is imported.

Reply:
> If easily possible, otherwise just everytime some device array is imported.

The current version of the PR already does it only when importing a device array (in Python). 
But we have the struct pointer, so it should be easy to just check `ArrowDeviceArray.device_type == 2` before calling the C++ Import function

Reply:
@pitrou  does the latest version with a lazy check (only when actually importing data, and only when the data has a CUDA device) look OK?